### PR TITLE
LibWeb: refactor StackingContext::paint_internal

### PR DESCRIPTION
**Pull request by maddanio**
This makes StackingContext::paint_internal easier to read and understand

### CODE REVIEW

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 1]
```diff
@@ -130,10 +130,18 @@ void StackingContext::paint_descendants(PaintContext& context, Layout::Node cons
 
 void StackingContext::paint_internal(PaintContext& context) const
 {
-    // For a more elaborate description of the algorithm, see CSS 2.1 Appendix E
-    // Draw the background and borders for the context root (steps 1, 2)
-    paint_node(m_box, context, PaintPhase::Background);
-    paint_node(m_box, context, PaintPhase::Border);
+    // helpers
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:133]: This comment (ironically) is not very helpful so I think we don't need it

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 2]
```diff
@@ -130,10 +130,18 @@ void StackingContext::paint_descendants(PaintContext& context, Layout::Node cons
...
+
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:134]: We usually don't have an empty line to start a function

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 3]
```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
             paintable->after_children_paint(context, PaintPhase::Foreground, should_clip_overflow);
     };
 
+    // For a more elaborate description of the algorithm, see CSS 2.1 Appendix E
+    // Draw the background and borders for the context root (steps 1, 2)
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:158]: I think this comment should be at the top of the function still, that's where we usually put comments like this

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 4]
```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
...
+    };
+
+    paint_box_content(m_box);
 
     // Draw positioned descendants with z-index `0` or `auto` in tree order. (step 8)
     // NOTE: Non-positioned descendants that establish stacking contexts with z-index `0` or `auto` are also painted here.
     // FIXME: There's more to this step that we have yet to understand and implement.
-    m_box.paint_box()->for_each_in_subtree_of_type<PaintableBox>([&](PaintableBox const& paint_box) {
-        auto const& z_index = paint_box.computed_values().z_index();
+    paintable().for_each_in_subtree_of_type<PaintableBox>([&](PaintableBox const& paint_box) {
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:183]: Why use `paintable()` here? Everything else still uses `m_box`

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 5]
```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
...
+        auto z_index = paint_box.computed_values().z_index();
         if (auto* child = paint_box.stacking_context()) {
-            if (!z_index.has_value() || z_index.value() == 0)
+            if (z_index.value_or(0) == 0)
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:186]: I personally find the previous way this was written easier to understand.
But feel free not to listen to me if other disagree here

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 6]
```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
...
                 paint_child(child);
             return TraversalDecision::SkipChildrenAndContinue;
-        }
-        if (z_index.has_value() && z_index.value() != 0)
-            return TraversalDecision::Continue;
-        if (!paint_box.layout_box().is_positioned())
+        } else {
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:189]: Why use an `else` here if the upper branch always returns?

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp [hunk 7]
```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
...
-        paint_descendants(context, paint_box.layout_box(), StackingContextPaintPhase::FocusAndOverlay);
-
-        return TraversalDecision::Continue;
+        }
     });
 
     // Draw other positioned descendants (step 9)
     for (auto* child : m_children) {
-        if (child->m_box.computed_values().z_index().has_value() && child->m_box.computed_values().z_index().value() >= 1)
+        if (child->m_box.computed_values().z_index().has_value() && child->m_box.computed_values().z_index().value() > 0)
```
Comment (davidot) [inline Userland/Libraries/LibWeb/Painting/StackingContext.cpp:201]: The spec says: `Stacking contexts formed by positioned descendants with z-indices greater than or equal to 1 in z-index order (smallest first) then tree order.`

We try to match the spec language as close as possible so I think we should keep the `>= 1`

(Unless I found an old spec)

**Review by davidot — CHANGES_REQUESTED** [review]
Hi, welcome to Serenity!

Always good to refactor the code, and it resolves that very new FIXME.
I mostly said some stuff about the style I don't have (much) experience in LibWeb painting. So I also don't know how logical the `paint_box_background` name is since it also draws the border.

Have you tested this with a couple of websites as a smoke test?

And finally, because I know Linus will say this otherwise, please reword your commit not to just say refactor. 
See https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy, 
> Don't Use weasel-words like "refactor" or "fix" to avoid explaining what's being changed.

Something like `LibWeb: Extract common paint operations to lambdas` might be something in the right direction

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 4 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:183]: actually no, hit_test uses this, it looks shorter, and I just wanred to align those two a bit more

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 5 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:186]: hmm, it is much shorter to mentally parse, and this underlines more the fact that the text treats 0 and auto pretty much the same at this level

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 1 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:133]: removed it

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 3 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:158]: moved

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 2 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:134]: removed it

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 6 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:189]: I find if/else much easier to parse to be honest, i parse them top-level as two branches of a decision, but I am not strong about this

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 7 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:201]: hmm, yes, well, when reading the code you basically have three classes: negative, 0/auto, positive, and I found it much clearer this way that this is the case. i find the spec wording unnneccessarily confusing when it clearly sorts things into three buckets, but uses different formulations for the buckets. but again, not stron about it

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 4 above)
Comment (davidot) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp:183]: I meant in this function, but maybe it's just my inexperience with the code

**Comment by maddanio** [comment]
sorry for abusing the CI to fix my code...its gotten too late. will try not to be that lazy next time :). finally builds now.

File: Userland/Libraries/LibWeb/Painting/StackingContext.cpp (hunk 7 above)
Comment (maddanio) [reply Userland/Libraries/LibWeb/Painting/StackingContext.cpp]: but I do see your point..changed it back :)

**Pull request closed without merging**